scalar reconfigure: help users remove buggy repos

derrickstolee opened the pull request:
When running 'scalar reconfigure -a', such as at install time, Scalar
has warning messages about the repository missing (or not containing a
.git directory). Failures can also happen while trying to modify the
repository-local config for that repository.

These warnings may seem confusing to users who don't understand what
they mean or how to stop them.

Add a warning that instructs the user how to remove the warning in
future installations.

derrickstolee committed:
scalar reconfigure: help users remove buggy repos

When running 'scalar reconfigure -a', such as at install time, Scalar
has warning messages about the repository missing (or not containing a
.git directory). Failures can also happen while trying to modify the
repository-local config for that repository.

These warnings may seem confusing to users who don't understand what
they mean or how to stop them.

Add a warning that instructs the user how to remove the warning in
future installations.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

dscho commented on contrib/scalar/scalar.c:
```diff
@@ -1368,7 +1369,12 @@ static int cmd_reconfigure(int argc, const char **argv)
+
+		if (failed) {
+			res = failed;
+			warning_errno(_("to unregister this repository from Scalar, run 'git config --global --unset --fixed-value scalar.repo \"%s\"'"), dir);
```
Could we have a line break before the command? Or even better, an empty line and a tab to indent it.

derrickstolee replied:
Interesting. Do we have precedent for that? The newlines don't get prefixed with "warning: ", which is my only concern.

dscho replied:
`git grep warn.*\\\\n` comes up with a few example, e.g. https://github.com/git/git/blob/v2.36.1/attr.c#L403

dscho reviewed (approved):
Thanks!

Just a minor suggestion about the format of the message.

BTW there's also talk over in Git for Windows [that the quoting might be incorrect for PowerShell/CMD and how to address that](https://github.com/git-for-windows/git/issues/3856#issuecomment-1129104434), but nothing conclusive came out of it. Once it does (if it does), we will want to touch this error message, too.

derrickstolee committed:
setup: add discover_git_directory_reason()

There are many reasons why discovering a Git directory may fail. In
particular, 8959555cee7 (setup_git_directory(): add an owner check for
the top-level directory, 2022-03-02) added ownership checks as a
security precaution.

Callers attempting to set up a Git directory may want to inform the user
about the reason for the failure. For that, expose the enum
discovery_result from within setup.c and into cache.h where
discover_git_directory() is defined.

I initially wanted to change the return type of discover_git_directory()
to be this enum, but several callers rely upon the "zero means success".
The two problems with this are:

1. The zero value of the enum is actually GIT_DIR_NONE, so nonpositive
   results are errors.

2. There are multiple successful states, so some positive results are
   successful.

Instead of updating all callers immediately, add a new method,
discover_git_directory_reason(), and convert discover_git_directory() to
be a thin shim on top of it.

Because there are extra checks that discover_git_directory_reason() does
after setup_git_directory_gently_1(), there are other modes that can be
returned for failure states. Add these modes to the enum, but be sure to
explicitly add them as BUG() states in the switch of
setup_git_directory_gently().

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

derrickstolee commented on contrib/scalar/scalar.c:
```diff
@@ -1349,26 +1349,48 @@ static int cmd_reconfigure(int argc, const char **argv)
+			goto loop_end;
+		}
+
+		discover_result = discover_git_directory_reason(&commondir, &gitdir);
+
+		switch (discover_result) {
```
Now that I look at it here in the diff, I could just put the method call in the switch.

dscho reviewed (approved):
Very nice!

dscho commented:
In the meantime, I integrated #509 so that the CI should pass again.
